syntax = "proto3";

package streaming_services;

message PushTextRequest {
  string text = 1;
  string time_stamp = 2;
}

message PushTextResponse {
  string request_status = 1;
}

message PopTextRequest {
  string time_stamp = 1;
}

message PopTextResponse {
  string text = 1;
  string request_status = 2;
}

message MatrixRow {
  repeated float elements = 1;
}

message Matrix {
  repeated MatrixRow rows = 1;
}

message MatrixList {
  repeated Matrix matrix = 1;
}

message PushFrameRequest {
  MatrixList frame = 1;
  string time_stamp = 2;
}

message PushFrameResponse {
  string request_status = 1;
}

message PopFrameRequest {
  string time_stamp = 1;
}

message PopFrameResponse {
  MatrixList frame = 1;
  string request_status = 2;
}

message PushImageRequest {
  bytes image = 1;
  string time_stamp = 2;
}

message PushImageResponse {
  string request_status = 1;
}

message PopImageRequest {
  string time_stamp = 1;
}

message PopImageResponse {
  bytes image = 1;
  string request_status = 2;
}

service Streaming {
  rpc PushText(PushTextRequest) returns (PushTextResponse) {}
  rpc PopText(PopTextRequest) returns (stream PopTextResponse) {}
  rpc PushFrame(PushFrameRequest) returns (PushFrameResponse) {}
  rpc PopFrame(PopFrameRequest) returns (stream PopFrameResponse) {}
  rpc PushImage(PushImageRequest) returns (PushImageResponse) {}
  rpc PopImage(PopImageRequest) returns (stream PopImageResponse) {}
}